cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -442,9 +440,10 @@ private MetadataSearchResponse search(Set<MetadataScope> scopes, String namespac
     Map<NamespacedEntityId, Metadata> systemMetadata = fetchMetadata(sortedEntities, MetadataScope.SYSTEM);
     Map<NamespacedEntityId, Metadata> userMetadata = fetchMetadata(sortedEntities, MetadataScope.USER);
 
+    Set<MetadataSearchResultRecord> resultRecords = addMetadataToEntities(sortedEntities, systemMetadata, userMetadata);
     return new MetadataSearchResponse(
-      sortInfo.getSortBy() + " " + sortInfo.getSortOrder(), offset, limit, numCursors, total,
-      addMetadataToEntities(sortedEntities, systemMetadata, userMetadata), cursors, showHidden
+      sortInfo.getSortBy() + " " + sortInfo.getSortOrder(), offset, limit, numCursors, resultRecords.size(),
```

Comment:
Currently, without the fix, the query will return a correct total for any offset given that the `sort` param is not given. If the `sort` param is given, the total is incorrect without the fix.

I am currently writing a new patch to fix this because I think I've found the root of the problem.